query/src/main/scala/filodb/query/LogicalPlan.scala

```diff
@@ -155,6 +155,68 @@ case class PeriodicSeries(rawSeries: RawSeriesLikePlan,
+ *  is wider. query_range can be called with start/end/step parameters only on the
+ *  top most expression in the query while subqueries can be:
+ *  (1) top level expression (complete equivalent to query_range)
+ *  (2) arguments to time range functions, potentially nested several levels deep
+ *
+ *  So, there are two major cases for a subquery expression:
+ *  A) any_instant_expression[W:S] as a top level expression. In this case, subquery
+ *     would issue "any_instant_expression" with its own start, end, and step
+ *     parameters. If such an expression is called with query_range API, where start != end
+ *     and step is not zero, an exception is thrown. No special subquery
+ *     node in logical plan is generated as all of the PeriodicSeries logical plans can
+ *     handle range_query API's start, step, and end parameters.
+ *  B) RangeFunction(any_instant_expression[W:S]) at any node/level of abstract syntax tree.
+ *     In this case, we ALWAYS have a range function involved and SubqueryWithWindowing
+ *     logical plan node is generated. The plan is almost identical in the functionality to
+ *     PeriodicSeriesWithWindowing. The difference between the two is that currently
+ *     PeriodicSeriesWithWindowing expects a RawSeriesLikePlan while subquery is NOT
+ *     necessarily an instant selector. PeriodicSeriesWithWindowing needs to be refactored,
+ *     so, it could handle any PeriodicSeries; however, at this point we are going to
+ *     replicate the functionality in the specialized SubqueryWithWindowing logical plan.
+ *     We do so, that we can integrage subquery code without affecting existing existing
+ *     code paths, stabilize the codebase, and later merge the two.
+ *  Below is an example of B) case:
+ *  sum_over_time(<someExpression>[5m:1m]) called with query_range API parameters:
+ *  start=S, end=E, step=ST
+ *  The above is equivalent to query API call:
+ *  sum_over_time(<someExpression[5m:1m])[(now-S):ST] offset (now-E)
+ */
+case class SubqueryWithWindowing(
+  innerPeriodicSeries: PeriodicSeriesPlan, // someExpression
+  startMs: Long, // S
+  stepMs: Long, // ST
+  endMs: Long, // E
+  functionId: RangeFunctionId, // sum_over_time
+  functionArgs: Seq[FunctionArgsPlan] = Nil,
+  subqueryWindowMs: Long, // 5m
+  subqueryStepMs: Long //1m
+) extends PeriodicSeriesPlan with NonLeafLogicalPlan {
+  override def children: Seq[LogicalPlan] = Seq(innerPeriodicSeries)
+  override def replacePeriodicSeriesFilters(filters: Seq[ColumnFilter]): PeriodicSeriesPlan = ???
+}
+
+case class TopLevelSubquery(
+  innerPeriodicSeries: PeriodicSeriesPlan, // someExpression
+  startMs: Long,
+  stepMs: Long,
+  endMs: Long,
+  subqueryStepMs: Long
+) extends PeriodicSeriesPlan with NonLeafLogicalPlan {
+  override def children: Seq[LogicalPlan] = Seq(innerPeriodicSeries)
+  override def replacePeriodicSeriesFilters(filters: Seq[ColumnFilter]): PeriodicSeriesPlan = ???
```

Comment:
Needs to be completed for long time range planner spanning raw and downsample queries